lib/std/fmt.zig

```diff
@@ -977,6 +988,11 @@ pub fn formatBuf(
     options: FormatOptions,
     writer: anytype,
 ) !void {
+    var fill_buffer: [4]u8 = undefined;
+    const fill_utf8 = if (std.unicode.utf8Encode(options.fill, &fill_buffer)) |len|
+        fill_buffer[0..len]
+    else |_|
+        " ";
```

Comment:
My bad, I didn't intentionally put it anywhere else than you suggested.

> Also, after looking at the implementation of formatUnicodeCodepoint above, it seems like this should actually be something more ...

So you suggest to handle the error as before with `fn utf()`? Are you sure about `else |err| switch (err)` vs `catch |err| switch (err)`? Being new to zig I find it surprising that I'm able to hide the error in an if-assign-expression, without the compiler telling me. Of course errors are just return values but isn't that the point that errors must be caught or bubble?

Reply:
> So you suggest to handle the error as before with `fn utf()`?

I'm suggesting that instead of replacing invalid characters with the default fill character of a space, they really ought to be replaced with the unicode replacement character (`unicode.replacement_character`) as is done in `fmt.FormatUnicodeCodepoint` instead of hiding the invalid input with default behavior.

> Are you sure about `else |err| switch (err)` vs `catch |err| switch (err)`? Being new to zig I find it surprising that I'm able to hide the error in an if-assign-expression, without the compiler telling me. Of course errors are just return values but isn't that the point that errors must be caught or bubble?

`x catch |_| y` produces the non-error payload of `x` if `x` is not an error, and `y` if `x` is an error.
`if (x) |_| y else |_| z` produces `y` if `x` is not an error and `z` if `x` is an error.
Note that in this case, in the case of no error, we don't want the error union payload directly (`len`) but some expression depending on it (`fill_buffer[0..len]`) which is the only reason to use `if () |_| else |_|` over `catch`.  In both cases the error is being "caught", just with a different syntax.  Also note that with this syntax, the `else` is mandatory, since as you said, errors cannot be ignored.

Reply:
cool thx for taking the time to explain.

Any note on my EDIT on `formatUnicodeCodepoint`
> However this function also only gets called once in the whole project and seems like the same problem like my utf function from before. As a utility it seems better be located in unicode package

Reply:
The `formatUnicodeCodepoint` function passes slices which point to a local to function which do not capture these values, and so these references are perfectly fine until the end of the scope where the local is declared.  The specific issue with the `utf` function was returning these slices outside of the scope containing the local.  I don't agree that `formatUnicodeCodepoint` belongs in the `unicode` namespace because it is a public function for formatting something, and should go with all of the other functions for formatting other things.  It already uses `unicode.utf8Encode` to perform the unicode-specific logic, although of course, it could now be cleaned up by using the new `unicode.utf8EncodeComptime` function if you would like to make that change.

Reply:
>  although of course, it could now be cleaned up by using the new unicode.utf8EncodeComptime function if you would like to make that change.

Yep I was also referring to DRY it up moving the utf part to unicode.

Is `formatUnicodeCodepoint` purely used at comptime like in this one occurrence in fmt? That might be a breaking change to a pub fn.

Reply:
> Yep I was also referring to DRY it up moving the utf part to unicode.

I'm saying that the unicode parts are already in unicode.  You could imagine situations where `error.Utf8CannotEncodeSurrogateHalf` should be replaced with the replacement character but `error.CodepointTooLarge` should be propagated to the caller, for example, or be considered programmer error.  I consider this specific behavior to be a choice being made by the formatting code, and so the logic should be in `fmt.zig`.

> Is `formatUnicodeCodepoint` purely used at comptime like in this one occurrence in fmt? That might be a breaking change to a pub fn.

I would just replace the inner call, the outer call should stay the same since it could be runtime-known as you say.